sticker/src/tensorflow/dataset.rs

```diff
@@ -209,3 +231,61 @@ where
         None
     }
 }
+
+/// An Iterator adapter performing local shuffling.
+///
+/// Fills a buffer with size `buffer_size` on the first call. Subsequent
+/// calls swap the next incoming item with a random element from the
+/// buffer and return the random element.
+pub struct Shuffled<I> {
+    inner: I,
+    buffer: Vec<Result<Sentence, Error>>,
+    buffer_size: usize,
+    shuffler: XorShiftRng,
+}
+
+impl<I> Iterator for Shuffled<I>
+where
+    I: Iterator<Item = Result<Sentence, Error>>,
+{
+    type Item = Result<Sentence, Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.buffer.is_empty() {
+            // fill the buffer
+            while let Some(sent) = self.inner.next() {
+                self.buffer.push(sent);
+                if self.buffer.len() == self.buffer_size {
+                    break;
+                }
+            }
+        }
+
+        let sent = match self.inner.next() {
+            Some(mut sent) => {
+                let buffer_len = self.buffer.len();
+                std::mem::swap(
```

Comment:
This seems to be biased in two ways:

1. It should (probably?) be possible to randomly draw the new element (`sent` here).
2. Using the remainder operator to index into a range introduces bias. E.g., consider a buffer of length 2 and a (hypothetical) random number generator that produces numbers `[0,2]`. Index `0` has twice the probability of being drawn as index `1`. Use [`Uniform`](https://docs.rs/rand/0.7.2/rand/distributions/uniform) or [`Rng::gen_range`](https://docs.rs/rand/0.7.2/rand/trait.Rng.html#method.gen_range).

Reply:
Not sure which one to use here. We are sampling quite often and the range stays the same until `inner` is exhausted, which probably means `Uniform` would be the best choice. OTOH, we'd need to take care of the possible out-of-bounds that can occur if we continue to sample from `Uniform` after `inner` is exhausted.

Reply:
`Rng::gen_range` should be ok. It calls this implementation:

https://docs.rs/rand/0.7.2/src/rand/distributions/uniform.rs.html#420

The initialization part is really cheap:  a subtraction to get a range, then: count leading zeros, bit shift, subtraction (for which x86_64 has instructions) to determine the largest multiple of the range.